core/src/main/java/org/opensearch/sql/planner/physical/MLCommonsOperator.java

```diff
@@ -0,0 +1,155 @@
+            .build();
+    MLPredictionOutput predictionResult = (MLPredictionOutput) machineLearningClient
+            .trainAndPredict(mlinput)
+            .actionGet(30, TimeUnit.SECONDS);
+    Iterator<Row> inputRowIter = inputDataFrame.iterator();
+    Iterator<Row> resultRowIter = predictionResult.getPredictionResult().iterator();
+    iterator = new Iterator<ExprValue>() {
+      @Override
+      public boolean hasNext() {
+          return inputRowIter.hasNext();
+      }
+
+      @Override
+      public ExprValue next() {
+        ImmutableMap.Builder<String, ExprValue> resultBuilder = new ImmutableMap.Builder<>();
+        resultBuilder.putAll(convertRowIntoExprValue(inputDataFrame.columnMetas(),
+                inputRowIter.next()));
+        resultBuilder.putAll(convertRowIntoExprValue(
+                predictionResult.getPredictionResult().columnMetas(),
+                resultRowIter.next()));
+        return ExprTupleValue.fromExprValueMap(resultBuilder.build());
+      }
+    };
+  }
+
+  @Override
+  public <R, C> R accept(PhysicalPlanNodeVisitor<R, C> visitor, C context) {
+    return visitor.visitMLCommons(this, context);
+  }
+
+  @Override
+  public boolean hasNext() {
+    return iterator.hasNext();
+  }
+
+  @Override
+  public ExprValue next() {
+    return iterator.next();
+  }
+
+  @Override
+  public List<PhysicalPlan> getChild() {
+    return Collections.singletonList(input);
+  }
+
+  protected MLAlgoParams convertArgumentToMLParameter(Argument argument, String algorithm) {
+    switch (FunctionName.valueOf(algorithm.toUpperCase())) {
+      case KMEANS:
+        return KMeansParams.builder().centroids((Integer) argument.getValue().getValue()).build();
+
+      default:
+        throw new IllegalArgumentException("unsupported argument type:"
+                + argument.getValue().getType());
+    }
+  }
+
+  private Map<String, ExprValue> convertRowIntoExprValue(ColumnMeta[] columnMetas, Row row) {
+    ImmutableMap.Builder<String, ExprValue> resultBuilder = new ImmutableMap.Builder<>();
+    for (int i = 0; i < columnMetas.length; i++) {
+      ColumnValue columnValue = row.getValue(i);
+      String resultKeyName = columnMetas[i].getName();
+      switch (columnValue.columnType()) {
+        case INTEGER:
+          resultBuilder.put(resultKeyName, new ExprIntegerValue(columnValue.intValue()));
+          break;
+        case DOUBLE:
+          resultBuilder.put(resultKeyName, new ExprDoubleValue(columnValue.doubleValue()));
+          break;
+        case STRING:
+          resultBuilder.put(resultKeyName, new ExprStringValue(columnValue.stringValue()));
+          break;
+        default:
+          break;
+      }
+    }
+    return resultBuilder.build();
+  }
+
+  private DataFrame generateInputDataset() {
+    List<Map<String, Object>> inputData = new LinkedList<>();
+    while (input.hasNext()) {
+      Map<String, Object> items = new HashMap<>();
+      input.next().tupleValue().forEach((key, value) -> {
```

Comment:
MLCommons doesn't accept non-numerical data to run Kmeans. How about check if the value is numerical or not first? Maybe just choose numerical columns to pass to KMeans. Not so urgent, you can do this in next PR. Add some todo now is ok.

Reply:
Added a check to only pass numerical columns to Kmeans in the next revision